pallene/to_ir.lua

```diff
@@ -205,6 +205,71 @@ function ToIR:convert_stat(cmds, stat)
 
         table.insert(cmds, ir.Cmd.For(stat.loc, v, start, limit, step, ir.Cmd.Seq(body)))
 
+    elseif tag == "ast.Stat.ForIn" then
+        local decls = stat.decls
+        local exps = stat.exps
+
+        -- for-in loops are desugared into regurlar loops before compiling.
+        -- For example, a loop like this:
+        --- ```
+        -- for a: T, b: T2 in <RHS> do
+        --     <loop body>
+        -- end
+        ---```
+        -- is compiled as if the following was written instead:
+        --```
+        -- local iter, st, ctrl = RHS[1], RHS[2], RHS[3]
+        -- while true do
+        --   local a_any, b_any = iter(st, ctrl)
+        --   if a_any == nil then break end
+        --   local a = a_any as T
+        --   local b = b_any as T2
+        --   ctrl = i_any
+        --   <loop body>
+        -- end
+        -- ```
```

Comment:
A question: Would it be a good idea to unify the `a_any` and `ctrl` variables, to create one less variable?

```
local iter, st, a_any, b_any
iter, st, a_any = RHS[1], RHS[2], RHS[3]
while true do
    a_any, b_any = iter(st, a_any)
    if a_any = nil then break end
    local a = a_any as T1
    local b = b_any as T2
    <loop body>
end
```

Reply:
Not sure, My personal POV is that these are locals anyway and not using the same symbol would make it much easier to spot the 1:1 correspondence between variables inserted by the compiler and those written by the user. That said, I think it's best we go with your gut. What do you say?